### PR TITLE
Fix a stream lifecycle bug in CustomDeviceLogReader

### LINKED ISSUE #94309: flutter run throws BadState: Cannot add new events after calling close
## Steps to Reproduce

1. Go to the stable channel
2. In the flutter SDK directory, checkout `origin/stable` or `upstream/stable` (otherwise you can't enable the custom-devices feature)
3. `flutter config --enable-custom-devices`
4. `flutter custom-devices add` and add some custom device
5. Preferably with an embedder that outputs a bunch of data to stdout
6. `flutter run -d my-custom-device`
7. Bad State: `logLines

The bug has been there since the beginning, but it's only present when a lot of stuff is logged from the embedder (or from the forwardPortCommand, not sure). (The exception happens when a new line is added to the `logLines` stream of `CustomDeviceLogReader`)

## command

`flutter run -d pi`

## exception

StateError: Bad state: Cannot add new events after calling close

```
#0      _BroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:243:24)
#1      _rootRunUnary (dart:async/zone.dart:1436:47)
#2      _CustomZone.runUnary (dart:async/zone.dart:1335:19)
#3      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1244:7)
#4      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#5      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#6      _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:63:11)
#7      _EventSinkWrapper.add (dart:async/stream_transformers.dart:13:11)
#8      _StringAdapterSink.add (dart:convert/string_conversion.dart:238:11)
#9      _LineSplitterSink._addLines (dart:convert/line_splitter.dart:151:13)
#10     _LineSplitterSink.addSlice (dart:convert/line_splitter.dart:126:5)
#11     StringConversionSinkMixin.add (dart:convert/string_conversion.dart:163:5)
#12     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:111:24)
#13     _rootRunUnary (dart:async/zone.dart:1436:47)
#14     _CustomZone.runUnary (dart:async/zone.dart:1335:19)
#15     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1244:7)
#16     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#17     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#18     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:63:11)
#19     _EventSinkWrapper.add (dart:async/stream_transformers.dart:13:11)
#20     _WindowsCodePageDecoderSink.add (dart:io/string_transformer.dart:124:11)
#21     _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:72:18)
#22     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:111:24)
#23     _rootRunUnary (dart:async/zone.dart:1436:47)
#24     _CustomZone.runUnary (dart:async/zone.dart:1335:19)
#25     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1244:7)
#26     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#27     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#28     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#29     _StreamController._add (dart:async/stream_controller.dart:607:7)
#30     _StreamController.add (dart:async/stream_controller.dart:554:5)
#31     _Socket._onData (dart:io-patch/socket_patch.dart:2166:41)
#32     _rootRunUnary (dart:async/zone.dart:1444:13)
#33     _CustomZone.runUnary (dart:async/zone.dart:1335:19)
#34     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1244:7)
#35     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#36     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#37     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#38     _StreamController._add (dart:async/stream_controller.dart:607:7)
#39     _StreamController.add (dart:async/stream_controller.dart:554:5)
#40     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1702:33)
#41     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1213:14)
#42     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#43     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#44     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:120:13)
#45     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:185:5)
```

## flutter doctor

```
[Ô£ô] Flutter (Channel unknown, 2.5.3, on Microsoft Windows [Version 10.0.19043.1348], locale de-DE)
    ÔÇó Flutter version 2.5.3 at C:\Users\hanne\devel\flutter_sdk
    ÔÇó Upstream repository unknown
    ÔÇó Framework revision 18116933e7 (6 weeks ago), 2021-10-15 10:46:35 -0700
    ÔÇó Engine revision d3ea636dc5
    ÔÇó Dart version 2.14.4

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    ÔÇó Android SDK at C:\Users\hanne\AppData\Local\Android\sdk
    Ô£ù cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    Ô£ù Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[Ô£ô] Chrome - develop for the web
    ÔÇó Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[Ô£ô] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.1)
    ÔÇó Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    ÔÇó Visual Studio Community 2019 version 16.11.31613.86
    ÔÇó Windows 10 SDK version 10.0.19041.0

[Ô£ô] Android Studio (version 2020.3)
    ÔÇó Android Studio at C:\Program Files\Android\Android Studio
    ÔÇó Flutter plugin can be installed from:
      ­ƒö¿ https://plugins.jetbrains.com/plugin/9212-flutter
    ÔÇó Dart plugin can be installed from:
      ­ƒö¿ https://plugins.jetbrains.com/plugin/6351-dart
    ÔÇó Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[Ô£ô] IntelliJ IDEA Ultimate Edition (version 2020.3)
    ÔÇó IntelliJ at C:\Users\hanne\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\203.7148.57
    ÔÇó Flutter plugin can be installed from:
      ­ƒö¿ https://plugins.jetbrains.com/plugin/9212-flutter
    ÔÇó Dart plugin can be installed from:
      ­ƒö¿ https://plugins.jetbrains.com/plugin/6351-dart

[Ô£ô] IntelliJ IDEA Ultimate Edition (version 2021.1)
    ÔÇó IntelliJ at C:\Users\hanne\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\211.7142.45
    ÔÇó Flutter plugin can be installed from:
      ­ƒö¿ https://plugins.jetbrains.com/plugin/9212-flutter
    ÔÇó Dart plugin can be installed from:
      ­ƒö¿ https://plugins.jetbrains.com/plugin/6351-dart

[Ô£ô] VS Code (version 1.62.3)
    ÔÇó VS Code at C:\Users\hanne\AppData\Local\Programs\Microsoft VS Code
    ÔÇó Flutter extension version 3.28.0

[Ô£ô] Connected device (4 available)
    ÔÇó Windows (desktop)     ÔÇó windows ÔÇó windows-x64    ÔÇó Microsoft Windows [Version 10.0.19043.1348]
    ÔÇó Chrome (web)          ÔÇó chrome  ÔÇó web-javascript ÔÇó Google Chrome 96.0.4664.45
    ÔÇó Edge (web)            ÔÇó edge    ÔÇó web-javascript ÔÇó Microsoft Edge 96.0.1054.34
    ÔÇó Raspberry Pi (mobile) ÔÇó pi      ÔÇó linux-arm64    ÔÇó Raspberry Pi 4 Model B+

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by ardera**
- make CustomDeviceLogReader cancel any subscriptions before closing the logLines stream
- add a test
- fixes https://github.com/flutter/flutter/issues/94309

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/custom_devices/custom_device.dart [hunk 1]
```diff
@@ -83,18 +92,28 @@ class CustomDeviceLogReader extends DeviceLogReader {
...
-  void dispose() {
-    _logLinesController.close();
+  Future<void> dispose() async {
+    final List<Future<void>> futures = <Future<void>>[];
+
+    for (final StreamSubscription<String> subscription in subscriptions) {
+      futures.add(subscription.cancel());
+    }
+
+    futures.add(logLinesController.close());
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/custom_devices/custom_device.dart:110]: Should we not wait for the subscription cancels before closing?

Comment (ardera) [reply packages/flutter_tools/lib/src/custom_devices/custom_device.dart:110]: Is that a question or a statement? 😄 

Wasn't sure about that, the [docs](https://api.dart.dev/stable/2.14.4/dart-async/StreamSubscription/cancel.html) say the subscription no longer receives events after the call to cancel, however not sure if that means after the returned future completes or after the call returned.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/custom_devices/custom_device.dart:110]: Definitely a question :)

Ahh, that's interesting. I had to read it like three times, but it appears to stop receiving events after the initial call. LGTM

**Review by christopherfujino — APPROVED** [review]
LGTM, thanks for finding and fixing this one!

File: packages/flutter_tools/lib/src/custom_devices/custom_device.dart (hunk 1 above)
Comment (ardera) [reply packages/flutter_tools/lib/src/custom_devices/custom_device.dart:110]: Okay 👍

**Comment by ardera** [comment]
I mean I wrote that code so I'm fixing my own bug really, but yeah np :-)

**Pull request merged**